manager/aica_django/connectors/GraphDatabase.py

```diff
@@ -54,14 +69,147 @@ def dict_to_cypher(input_dict: dict[str, Any]) -> str:
+        super().__init__()
+        self.dropout = dropout
+        self.conv1 = SAGEConv(in_dim, hidden_dim, aggr=aggr)
+        self.conv2 = SAGEConv(hidden_dim, hidden_dim, aggr=aggr)
+        self.conv3 = SAGEConv(hidden_dim, out_dim, aggr=aggr)
+          
+    def forward(self, data: Data):
+
+        edge_idx = EdgeIndex(data.edge_index) 
+
+        x = self.conv1(data.x, edge_idx)
+        x = F.tanh(x)
+        x = F.dropout(x, p=self.dropout)
+        
+        x = self.conv2(x, edge_idx)
+        x = F.tanh(x)
+        x = F.dropout(x, p=self.dropout)
+        
+        x = self.conv3(x, edge_idx)
+        x = F.tanh(x)
+    
+        return x
+
+def load_graphml_data(BASEFILE, SHVVL_MAX_FEATURE_LEN=12, SHVVL_BANDWIDTH=20):
+    aica_graph = nx.read_graphml(BASEFILE)
+    node_ids = []
+    a = 0
+    for n in aica_graph.nodes(data=True):
+        typedata = n[1]["TYPE"]
+        node_ids.append(n[1]['identifier'])
+        del n[1]["identifier_vec"]
+        z = str(typedata) + "\0"
+        keys = sorted(n[1].keys())
+
+        for k in range(SHVVL_MAX_FEATURE_LEN):
+            z += str(n[1][keys[k]]) + "\0" if k < len(keys) else "\0"         
+        z = z[:-1]
+        shoveled_data = shvvl_float(z, SHVVL_BANDWIDTH)
+        shvvlsize = len(shoveled_data)
+        s : dict = {}
+
+        n[1].clear()
+        for x in range(shvvlsize):
+            pass
+            n[1]["SHVVL_ID" + str(x)] = shoveled_data[x]    
+        a = max(a, len(n[1]))
+        n[1]["TYPE"]=typedata
+    for x in aica_graph.edges(data=True):
+        x[2].clear()
+        x[2]["dummy"]=0
+
+    data_tensor = torch_geometric.utils.convert.from_networkx(aica_graph, ['SHVVL_ID' + str(x) for x in range(a)])
+    z = 0
+    for original_node in aica_graph.nodes(data=True):
+        if original_node[1]["SHVVL_ID0"] != data_tensor.x[z][0]:
+            raise BaseException("ERROR: Ordering not lined up")
+            z = -1
+            break
+        z += 1
+    if z != -1:
+        print("Likely ordered")
+
+    return data_tensor, node_ids
+    
+
+def run_graphsage(data_tensor, hidden_dim=128, in_dim=2080, out_dim=128): 
+    # make hidden_dim size be num_node_types*hidden_dim
+    model = AICASage(in_dim=in_dim, 
+                    hidden_dim=hidden_dim, 
+                    out_dim=out_dim)
+    
+    if torch.cuda.is_available():
+        print("Starting in GPU mode...")
+        deviceType =  f'cuda' 
+    else:
+        print("WARNING: NO VALID CUDA DEVICE FOUND! Defaulting to CPU...")
+        deviceType = 'cpu'
+
+    device = torch.device(deviceType)
+
+    model.to(device)
+    data_tensor.to(device)
+    emb = model(data_tensor) 
+    emb = emb.cpu().detach().numpy()
+
+    return emb
+
+def update_graph(emb, node_ids):
+    print("Updating knowledge graph ...")
+    graph_obj = AicaNeo4j(initialize_graph=False)
+    for i in range(emb.shape[0]):
+        if 'network-traffic' in node_ids[i]:
```

Comment:
probably doesn't matter but node_ids[i].startswith("network-traffic--") would be more precise.